### PR TITLE
`oauth`: Improved more granular error handling

### LINKED ISSUE #6261: [OAuth] Improve error handling
Type the results' errors of the connection interface with clear error codes and adjust status code based on that in the `oauth.rs` file + pass along the error code so that we can surface clearly actionable error codes in the OAuthAPIError `front`/`connectors` side

### PR DESCRIPTION
**Pull request by spolu**
## Description

Tie the `Connection` interface to always return a `ConnetionError` that we can use to return better errors codes and more precise codes and messages.

It's a bit verbose but much better in terms of error handling. Will be required to properly expose predictable and granular codes to front/connectors so that they can be properly actioned.

Fixes: https://github.com/dust-tt/dust/issues/6261

Also tweaks error_response used in `core` to use tracing::error! with a searchable message "API error".

From a devx perspective we can look for "API Error" and then filter on trace_span_id to get the whole log of the request and investigate logging of internal errors

## Risk

N/A

## Deploy Plan

- deploy `oauth`

### CODE REVIEW

File: core/src/oauth/connection.rs [hunk 1]
```diff
@@ -17,9 +17,45 @@ use serde::{Deserialize, Serialize};
...
 use std::time::Duration;
 use std::{env, fmt};
+use tracing::error;
 
 use super::providers::confluence::ConfluenceConnectionProvider;
 
+// Define the ErrorKind enum with serde attributes for serialization
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum ConnectionErrorCode {
```
Comment (fontanierh) [inline core/src/oauth/connection.rs:27]: `thiserror`  (that we use [here](https://github.com/dust-tt/dust/blob/0e46f244ddf7d43669505440b3a2d124c7e686e8/core/src/sqlite_workers/client.rs#L23)) is quite nice to work with

**Review by fontanierh — APPROVED** [review]
LGTM but might be nicer with `thiserror`

File: core/src/oauth/connection.rs (hunk 1 above)
Comment (spolu) [reply core/src/oauth/connection.rs:27]: Will move forward as is but no need to pass arguments that much so not sure we need it yet?

Comment (fontanierh) [reply core/src/oauth/connection.rs:27]: even without that, you can simply define variants and have the display automatically implemented (so no need to do one enum then one struct then one display impl)

**Pull request merged**